simplex/engine.go

```diff
@@ -0,0 +1,313 @@
+// Copyright (C) 2019-2025, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package simplex
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/ava-labs/simplex"
+	"go.uber.org/zap"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/proto/pb/p2p"
+	"github.com/ava-labs/avalanchego/snow"
+	"github.com/ava-labs/avalanchego/snow/engine/common"
+	"github.com/ava-labs/avalanchego/snow/engine/snowman/block"
+	"github.com/ava-labs/avalanchego/snow/validators"
+	"github.com/ava-labs/avalanchego/utils/logging"
+)
+
+var _ common.Engine = (*Engine)(nil)
+
+var (
+	maxProposalWaitTime = time.Second * 2
+	maxRebroadcastWait  = time.Second * 2
+	tickInterval        = time.Millisecond * 500
+
+	errUnknownMessageType = errors.New("unknown message type")
+	errConvertingMessage  = errors.New("error converting message to simplex message")
+)
+
+type Engine struct {
+	// nonValidator marks that this node is not a validator
+	// this is included, but marked as a todo since the e2e tests currently
+	// try and bootstrap a node that is not a validator(see func `CheckBootstrapIsPossible`).
+	// The bootstrapped node is a non-validator which is currently not supported.
+	// TODO: handle non-validators properly
+	nonValidator bool
+
+	// list of NoOpsHandler for messages dropped by engine
+	common.AllGetsServer
+	common.StateSummaryFrontierHandler
+	common.AcceptedStateSummaryHandler
+	common.AcceptedFrontierHandler
+	common.AcceptedHandler
+	common.AncestorsHandler
+	common.PutHandler
+	common.QueryHandler
+	common.ChitsHandler
+
+	// Handler that passes application messages to the VM
+	common.AppHandler
+	validators.Connector
+
+	epoch              *simplex.Epoch
+	blockDeserializer  *blockDeserializer
+	quorumDeserializer *QCDeserializer
+	logger             logging.Logger
+	vm                 block.ChainVM
+	consensusCtx       *snow.ConsensusContext
+}
+
+// THe VM must be initialized before creating the engine
+func NewEngine(consensusCtx *snow.ConsensusContext, ctx context.Context, config *Config) (*Engine, error) {
+	if _, ok := config.Validators[config.Ctx.NodeID]; !ok {
+		config.Log.Info("Node is not a validator for the subnet",
+			zap.Stringer("nodeID", config.Ctx.NodeID),
+			zap.Stringer("chainID", config.Ctx.ChainID),
+			zap.Stringer("subnetID", config.Ctx.SubnetID),
+		)
+		return nonValidatingEngine(consensusCtx, config)
+	}
+
+	signer, verifier := NewBLSAuth(config)
+	qcDeserializer := &QCDeserializer{
+		verifier: &verifier,
+	}
+
+	signatureAggregator := &SignatureAggregator{
+		verifier: &verifier,
+	}
+	comm, err := NewComm(config)
+	if err != nil {
+		return nil, err
+	}
+
+	storage, err := newStorage(ctx, config, qcDeserializer, &blockTracker{})
```

Comment:
we're passing an empty block tracker but then creating what seems to be the real block tracker down the line in line 105 and overriding it.

Maybe just create it earlier? 

are you doing that that because of the last block?